test/environments.js

```diff
@@ -0,0 +1,223 @@
+const utils = require('./utils');
+const fs = require('fs');
+const path = require('path');
+const rm = require('rimraf').sync;
+const {TestServer} = require('../utils/testserver/');
+
+class ServerEnvironment {
+  async beforeAll(state) {
+    const assetsPath = path.join(__dirname, 'assets');
+    const cachedPath = path.join(__dirname, 'assets', 'cached');
+
+    const port = 8907 + state.parallelIndex * 2;
+    state.server = await TestServer.create(assetsPath, port);
+    state.server.enableHTTPCache(cachedPath);
+    state.server.PORT = port;
+    state.server.PREFIX = `http://localhost:${port}`;
+    state.server.CROSS_PROCESS_PREFIX = `http://127.0.0.1:${port}`;
+    state.server.EMPTY_PAGE = `http://localhost:${port}/empty.html`;
+
+    const httpsPort = port + 1;
+    state.httpsServer = await TestServer.createHTTPS(assetsPath, httpsPort);
+    state.httpsServer.enableHTTPCache(cachedPath);
+    state.httpsServer.PORT = httpsPort;
+    state.httpsServer.PREFIX = `https://localhost:${httpsPort}`;
+    state.httpsServer.CROSS_PROCESS_PREFIX = `https://127.0.0.1:${httpsPort}`;
+    state.httpsServer.EMPTY_PAGE = `https://localhost:${httpsPort}/empty.html`;
+  }
+
+  async afterAll({server, httpsServer}) {
+    await Promise.all([
+      server.stop(),
+      httpsServer.stop(),
+    ]);
+  }
+
+  async beforeEach(state) {
+    state.server.reset();
+    state.httpsServer.reset();
+  }
+}
+
+class DefaultBrowserOptionsEnvironment {
+  constructor(defaultBrowserOptions, logger, playwrightPath) {
+    this._defaultBrowserOptions = defaultBrowserOptions;
+    this._logger = logger;
+    this._playwrightPath = playwrightPath;
+  }
+
+  async beforeAll(state) {
+    state.defaultBrowserOptions = {
+      ...this._defaultBrowserOptions,
+      logger: this._logger,
+    };
+    state.playwrightPath = this._playwrightPath;
+  }
+
+  async beforeEach(state, testRun) {
+    this._logger.setTestRun(testRun);
+  }
+
+  async afterEach(state) {
+    this._logger.setTestRun(null);
+  }
+}
+
+// simulate globalSetup per browserType that happens only once regardless of TestWorker.
```

Comment:
Let's use globalSetup!

Reply:
This will go away in the next patch - with first-class artifacts support in test runner.